ogr/ogrsf_frmts/sqlite/ogrsqlitelayer.cpp

```diff
@@ -3555,11 +3553,165 @@ OGRErr OGRSQLiteLayer::CommitTransaction()
 /************************************************************************/
 
 OGRErr OGRSQLiteLayer::RollbackTransaction()
-
 {
     return m_poDS->RollbackTransaction();
 }
 
+/************************************************************************/
+/*                          PrepareStartTransaction()                   */
+/************************************************************************/
+void OGRSQLiteLayer::PrepareStartTransaction()
+{
+    m_apoFieldDefnChanges.clear();
+    m_apoGeomFieldDefnChanges.clear();
+}
+
+/************************************************************************/
+/*                          FinishRollbackTransaction()                 */
+/************************************************************************/
+void OGRSQLiteLayer::FinishRollbackTransaction()
+{
+
+    // Deleted fields can be safely removed from the storage after being restored.
+    std::vector<int> toBeRemoved;
+    int idx{0};
+
+    // Loop through all changed fields and reset them to their previous state.
+    for (int i = static_cast<int>(m_apoFieldDefnChanges.size()) - 1; i >= 0;
+         i--)
+    {
+        auto &oFieldChange = m_apoFieldDefnChanges[i];
+        const char *pszName = oFieldChange.poFieldDefn->GetNameRef();
+        const int iField = oFieldChange.iField;
+        if (iField >= 0)
+        {
+            switch (oFieldChange.eChangeType)
+            {
+                case FieldChangeType::DELETE:
+                {
+                    GetLayerDefn()->AddFieldDefn(
+                        oFieldChange.poFieldDefn.get());
+
+                    // Now move the field to the right place
+                    // from the last position to its original position
+                    const int iFieldCount = GetLayerDefn()->GetFieldCount();
+                    int *panOrder =
```

Comment:
I would use a std::vector<int> . Use of CPLMalloc/CPLCalloc in this driver are old-school habits

Reply:
done